File: java/Notification.java
package smarthome;
import java.time.*;
public class Notification {
    private String message;
    private LocalDateTime timestamp;
    private String imagePath;
    private boolean read;
    public Notification(String message) { this.message = message; this.timestamp = LocalDateTime.now(); }
    public void send() {
        try {
            if (message == null || message.trim().isEmpty()) throw new IllegalArgumentException("Пустое сообщение");
            System.out.println("Уведомление: " + message);
        } catch (Exception e) {
            System.out.println("Ошибка при отправке уведомления: " + e.getMessage());
        }
    }
    public void show() { System.out.println(timestamp + ": " + message); }
    public void sendToDevice(Device d) {
        try {
            if (d == null) throw new IllegalArgumentException("Устройство null");
            d.perform("notify " + message);
        } catch (Exception e) { System.out.println("Ошибка при отправке уведомления устройству: " + e.getMessage()); }
    }
    public void repeat() { System.out.println("Повтор уведомления: " + message); }
    public void attachImage(String p) { imagePath = p; }
    public void markAsRead() { read = true; }
}

File: cpp/src/Notification.cpp
#include "Notification.hpp"

Notification::Notification(int deviceId, const std::string &msg): deviceId(deviceId), message(msg), timestamp(std::time(nullptr)) {}

Notification::Notification(const std::string &msg): deviceId(-1), message(msg), timestamp(std::time(nullptr)) {}
Notification::Notification(const Notification& other): deviceId(other.deviceId), message(other.message), timestamp(other.timestamp) {}

void Notification::sendToDevice(int deviceId) {
    std::cout << "Уведомление для устройства " << deviceId << ": " << message << std::endl;
}

void Notification::sendToAllUsers() {
    std::cout << "Широковещательное уведомление: " << message << std::endl;
}

void Notification::show() const {
    std::cout << "[Уведомление] " << message << " (время: " << std::asctime(std::localtime(&timestamp)) << ")" << std::endl;
}
